classes/class-wc-connect-payment-methods-store.php

```diff
@@ -72,10 +72,10 @@ protected function potentially_update_selected_payment_method_from_payment_metho
 				return;
 			}
 
-			// Has the stored method ID been removed? Select the first available one
+			// Has the stored method ID been removed, or is there only one available? Select the first available one
 			$selected_payment_method_id = $this->service_settings_store->get_selected_payment_method_id();
 			if (
-				$selected_payment_method_id &&
+				( $selected_payment_method_id || 1 === count( $payment_method_ids ) ) &&
```

Comment:
The requirements of this PR changed, so I think this code no longer works how we want it to.

From reading the PR conversation, we only want to set the payment method ID automatically if there is just one available payment method.

At this point we have the following information:

* `$payment_method_ids`: list of currently available payment method IDs
* `$selected_payment_method_id`: this the last selected payment ID. it may not exist in the list of payment IDs anymore (for example, if a user deleted it)

We want to select the payment ID automatically for the user in these two cases:
1. there was a payment ID selected before, but it's no longer available. there is only one available payment method
2. there is no currently selected payment method, and there is only one available.

For both of these we could check `1 === count( $payment_method_ids )` and if that's true, we can select the payment ID:

```
if ( 1 === count( $payment_method_ids ) {
  $this->service_settings_store->set_selected_payment_method_id( $payment_method_ids[ 0 ] );
}
```

Am I missing a scenario we want to cover?

Reply:
> For both of these we could check 1 === count( $payment_method_ids ) and if that's true, we can select the payment ID

Not quite. If there's a valid payment method selected, we don't want to do anything:
```php
if ( ! in_array( $selected_payment_method_id, $payment_method_ids ) && 1 === count( $payment_method_ids ) {
  $this->service_settings_store->set_selected_payment_method_id( $payment_method_ids[ 0 ] );
}
```

Otherwise I agree with your reasoning, and I can't think on any other scenario we're missing.

Reply:
The other scenario is the status quo in `master`: if there was already a selected payment method, we can just select the first stored one without worrying if it's the first out of multiple. This way, it never goes from selection to no selection unless there are no stored cards.

I interpreted the conversation above to refer to expanding the auto-selection to the case where there isn't an existing selection, but not restricting the condition for the case where there is one.

I'm not necessarily opposed to only changing to a new selection if there's no doubt in which card to select – do you all think that's best?

Reply:
> if there was already a selected payment method, we can just select the first stored one without worrying if it's the first out of multiple.

I don't understand this. Suppose we have these available payment methods: `[1234, 5678, 9012]`, and the selected method ID is `5678`. Why would you mark "the first one" as selected? That would reset it to `1234`.

Reply:
@DanReyLop Sorry, should have mentioned for completeness: as you mentioned, it would only change it if the method already selected is not available [[code](https://github.com/Automattic/woocommerce-services/blob/4a401a0a1482ed6f29245a79ea1f672cd8b9c5b7/classes/class-wc-connect-payment-methods-store.php#L78-L79)].

Reply:
> it would only change it if the method already selected is not available

Is this what we want, though?

That means we're blindly selecting a payment method on behalf of the user without any action on their part.

In fact I think it would doubly-strange because at some point this user had selected a different method that's no longer available. I don't see why they'd expect that when their previous method is no longer available, we'd choose a new one.

I did see that Dani mentioned this was probably ok because we say which credit card is being used before starting. I'm not so sure about this. Users who are not the credit card holders can still print labels (iirc) so that means that someone not familiar with the last 4 digits of the card may go through.

To play devil's advocate on my own arguments though: We've had things work this way for a while and heard no complaints :D. Maybe this means this does help our users by not asking them to choose a credit card every time there is more than one. Or maybe the number of users with multiple cards is small, so no one's run into a scenario they don't like.

In the off-chance this causes a problem there are steps to remediate:
0. update the selected credit card
1. re-charge the user using a BD manual payment for the amount of the receipts they don't like
2. refund the charges on the incorrect credit card

So, if people are leaning towards keeping the "choose a credit card for me out of many" behaviour, I've justified it to myself as well and let's go for it 👍

Reply:
> at some point this user had selected a different method that's no longer available

Is this to say we should reconsider a) auto-selecting one card out of multiple, b) auto-selecting a card when there was a previous selection, or c) auto-selecting a card in general?

Unless it's (c), I'd be in favor of going forward with this PR as is, since (at least from my perspective), it is scoped to expanding the auto-selection criteria to the no-card-selected case, not restricting the existing criteria – for which we could open another issue.

Reply:
> Is this to say we should reconsider a) auto-selecting one card out of multiple, b) auto-selecting a card when there was a previous selection, or c) auto-selecting a card in general?

In that specific phrase I was referring to (b).